### ISSUE TITLE
Incompatibility with lewis6991/fileline.nvim

### ISSUE THREAD

**Issue by pschmitt** [purarue/gitsigns-yadm.nvim#5]
Hi,

If anyone happens to use both this here plugin and lewis6991/fileline.nvim (or wsdjeg/vim-fetch) and then opens a git-tracked file like this: `nvim FILE:1` an error is thrown:

```
Error executing vim.schedule lua callback: ...cal/share/nvim/lazy/gitsigns.nvim/lua/gitsigns/async.lua:95: The async coroutine faile
d: ...al/share/nvim/lazy/gitsigns.nvim/lua/gitsigns/attach.lua:269: Invalid buffer id: 1
stack traceback:
        [C]: in function '__index'
        ...al/share/nvim/lazy/gitsigns.nvim/lua/gitsigns/attach.lua:269: in function 'fn'
        .../share/nvim/lazy/gitsigns.nvim/lua/gitsigns/debounce.lua:68: in function 'attach_throttled'
        ...al/share/nvim/lazy/gitsigns.nvim/lua/gitsigns/attach.lua:431: in function <...al/share/nvim/lazy/gitsigns.nvim/lua/gitsig
ns/attach.lua:430>
stack traceback:
        [C]: in function 'error'
        ...cal/share/nvim/lazy/gitsigns.nvim/lua/gitsigns/async.lua:95: in function 'cb'
        ...cal/share/nvim/lazy/gitsigns.nvim/lua/gitsigns/async.lua:145: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>

```

Sample config to reproduce:

```lua
for name, url in pairs({
  gitsigns = "https://github.com/lewis6991/gitsigns.nvim",
  -- ADD OTHER PLUGINS _NECESSARY_ TO REPRODUCE THE ISSUE
  gitsigns_yadm = "https://github.com/purarue/gitsigns-yadm.nvim",
  fileline = "https://github.com/lewis6991/fileline.nvim",
  plenary = "https://github.com/nvim-lua/plenary.nvim",
}) do
  local install_path = vim.fn.fnamemodify("gitsigns_issue/" .. name, ":p")
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system({ "git", "clone", "--depth=1", url, install_path })
  end
  vim.opt.runtimepath:append(install_path)
end

require("gitsigns").setup({
  debug_mode = true, -- You must add this to enable debug messages
  -- ADD GITSIGNS CONFIG THAT IS _NECESSARY_ FOR REPRODUCING THE ISSUE
  _on_attach_pre = function(_, callback)
    require("gitsigns-yadm").yadm_signs(callback)
  end,
})
```

and then: `nvim -u FILE_FROM_ABOVE.lua MYFILE:LINE_NUMBER` (eg: `nvim -u init-repro.lua README.md:1`)

**Comment by pschmitt** [comment]
Both the fileline and vim-fetch plugins seem to close the first buffer and re-open a new one when triggered, which is probably why this error is logged.

```
nvim /tmp/file
:buffers
  1 %a   "/tmp/file"                    line 1
```

vs:

```
nvim /tmp/file:1
:buffers
  2 %a   "/tmp/file"                    line 1
```

**Comment by purarue** [comment]
Thanks for reporting!

Just for my reference, need to actually open a file which is in the yadm repo for the error, so like a:

```
nvim -u init-repro.lua ~/.config/yadm/README.md:50
```

I think the error gets thrown in gitsigns code, not here, so probably cant catch that.

perhaps the argument can be checked to see if theres a line number included, I'm not sure what the API for that is like... seems quite hacky, but can take a look when I have some free time

**Comment by disrupted** [comment]
fwiw I am seeing the same error from gitsigns after making a Git commit using Neogit. I don't have the fileline.nvim plugin.

**Comment by purarue** [comment]
Just to confirm, are you using a line number when opening the file like `nvim README.md:1`?

**Comment by disrupted** [comment]
I always open files regularly, make some changes, and commit them using Neogit. It's only after the commit message is saved and the window closes that I see the error. It goes away when I disable gitsigns-yadm. fwiw I've never used the fileline plugin

**Comment by purarue** [comment]
Just to update the status here a bit, I am a bit unsure how to proceed here; this plugin is a bunch of configuration around [one specific callback command](https://github.com/purarue/gitsigns-yadm.nvim/blob/9813de8c122c62ce27a83a80e27c9b4fb662b018/lua/gitsigns-yadm/init.lua#L117-L120) which is the entire surface area of the gitsigns API this interacts with, it is not complicated at all.

Those two fields are then processed [here](https://github.com/lewis6991/gitsigns.nvim/blob/5f808b5e4fef30bd8aca1b803b4e555da07fc412/lua/gitsigns/attach.lua#L128-L145) in gitsigns.

I would have to read through/really understand gitsigns a lot more than I do currently to able to debug why this is failing, and thats not something I have a ton of time for right now.

If someone can point me in the right direction/would like to take this on and submit a PR, would be appreciated.

**Comment by disrupted** [comment]
I've temporarily fixed the issue for me by disabling this plugin for other projects (inside Git repos).

```lua
_on_attach_pre = function(bufnr, callback)
    if vim.uv.fs_stat '.git' then
        -- disable YADM if inside Git repo
        return callback()
    end
    require('gitsigns-yadm').yadm_signs(callback)
end
```

**Comment by purarue** [comment]
Hi, have finally had some time to look into this.

I've pushed some related fixes (skipping attach if a git directory is detected, keeping track of the bufnr and making sure its still valid when attaching), can you update and see if the errors persist?

**Comment by disrupted** [comment]
upon updating I am no longer seeing gitsigns in regular Git repositories. It also prints the message "Disabling inside git directory" on startup

**Comment by purarue** [comment]
Oh, forgot about that; removed the debug log (the 'disabling inside git directory' just meant disabling trying test if yadm could attach since we're in a git directory)

Are you using the standard/same configuration?

Does the error re-appear if you set `disable_inside_gitdir` to `false` in the config like described [here](https://github.com/purarue/gitsigns-yadm.nvim?tab=readme-ov-file#configuration)?

Perhaps the default for that should be false.

**Comment by purarue** [comment]
Oh, I dont think the config above would work anymore, since I don't call `callback` when its not detected as a yadm file, so you may need to do:

```lua
_on_attach_pre = function(bufnr, callback)
    if not vim.uv.fs_stat '.git' then
        require('gitsigns-yadm').yadm_signs(callback)
    end
end
```

**Comment by disrupted** [comment]
thanks, that works!

**Comment by purarue** [comment]
I think I was running into the same bug actually, re-added the `callback`s when disabling.

will close this since I think the original issue is gone, feel free to re-open or create a new issue if you're running into other issues

**Comment by disrupted** [comment]
this looks better, I no longer need the `.git` check.

However it seems the following workaround is still required to avoid attaching to `yadm commit` and `yadm rebase` buffers. Otherwise gitsigns shows a _added hunk_ for those buffers.

```lua
_on_attach_pre = function(bufnr, callback)
    local ignore_filetypes = {
        'gitcommit', -- YADM commit
        'gitrebase', -- YADM rebase
    }
    if
        vim.tbl_contains(
            ignore_filetypes,
            vim.bo[bufnr].filetype
        )
    then
        return -- do not attach to buffers of this filetype
    end
    require('gitsigns-yadm').yadm_signs(
        callback,
        { bufnr = bufnr }
    )
end
```

**Comment by pschmitt** [comment]
Erm I believe this went a bit too far. As of the current [HEAD](https://github.com/purarue/gitsigns-yadm.nvim/commit/7e2bfd891bff95651f65577a7f8e22276c0c46fd) when inside a regular git repo, I just get greeted by a notification stating "Disabling inside git directory" and gitsigns is disabled completely.

**Comment by purarue** [comment]
I think that was a few commits ago, could you try updating?

That notification is no longer in the code.

If gitsigns still isnt working, does it work if you pass the `disable_inside_gitdir = false` in the `opts` like

```lua
{
    "lewis6991/gitsigns.nvim",
    dependencies = {
        "nvim-lua/plenary.nvim",
        {
            "purarue/gitsigns-yadm.nvim",
            opts = {
                disable_inside_gitdir = false
            },
        },
    },
    opts = {
        _on_attach_pre = function(bufnr, callback)
            require("gitsigns-yadm").yadm_signs(callback, { bufnr = bufnr })
        end,
        -- other configuration for gitsigns...
    },
```

**Comment by pschmitt** [comment]
Classic user-error happened here: I didn't notice the change in `_on_attach_pre`.

All good now with `disable_inside_gitdir = true`. If set to `false` though, gitsigns just isn't available at all inside regular git repos. 
Is that really intended?

**Comment by purarue** [comment]
Hm, should probably not be happening...., can reproduce and yeah, that seems like a bug.

I think I'll add some debug logs tomorrow so its easier to track/report what is happening, should make it easier to make sure we're running into the same bug; will create a separate issue for that
